// @synapsejs/synapse/build/lib/index.js
"use strict";
/* eslint-disable import/extensions */
Object.defineProperty(exports, "__esModule", { value: true });
exports.synapse = void 0;
var http_1 = require("./protocol/http");
var sse_1 = require("./protocol/sse");
var ws_1 = require("./protocol/ws");
var utility_1 = require("./utility");
/** Initializes API request handlers from {@linkcode Resource} definitions in the given ```directory```.
 * @param directory A directory containing {@linkcode Resource} definitions.
 * @returns An object containing properties ```ws```, ```http```, and ```sse```, whose values are request handlers for the respective protocol.
 */
function synapse(directory, peers, pattern) {
    if (peers === void 0) { peers = []; }
    if (pattern === void 0) { pattern = []; }
    utility_1.requireAll(directory);
    var callback = utility_1.makeChain();
    return {
        http: http_1.default(callback),
        sse: sse_1.default(callback),
        ws: ws_1.default(callback, peers, pattern),
        use: callback.add,
    };
}
exports.synapse = synapse;
var State_1 = require("./control/State");
Object.defineProperty(exports, "State", { enumerable: true, get: function () { return State_1.default; } });
var Resource_1 = require("./state/Resource");
Object.defineProperty(exports, "Resource", { enumerable: true, get: function () { return Resource_1.default; } });
var Collection_1 = require("./state/Collection");
Object.defineProperty(exports, "Collection", { enumerable: true, get: function () { return Collection_1.default; } });
var Field_1 = require("./state/Field");
Object.defineProperty(exports, "Field", { enumerable: true, get: function () { return Field_1.default; } });
var Schema_1 = require("./state/Schema");
Object.defineProperty(exports, "Schema", { enumerable: true, get: function () { return Schema_1.default; } });
exports.fields = require("./fields");
exports.decorators = require("./@");
